### PR TITLE
Create SECURITY.md

### PR DESCRIPTION
**Pull request by GijsvanDulmen**
# Changes

Introduce a page with a single example of securing the dashboard. Feedback is needed on the example. It's rather highly specific for my case (assumes istio), but could maybe benefit others.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

### CODE REVIEW

**Comment by a-roberts** [comment]
Excellent, thanks for this @GijsvanDulmen!

Would be great for someone to give this a try as well, I certainly will as soon as I can. We can then look at moving platform specific security information in here too ((happy to do as follow-up, so this would be for the existing documentation that mentions a secure Ingress endpoint and OpenShift).

/ok-to-test

File: SECURITY.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# Securing the dashboard
+
+The following page describes some ways to secure the dashboard.
+
+## Basic Authentication on Istio with Envoy Filter
+When you are using Istio there is a way of introducing an envoy filter which contains
```
Comment (a-roberts) [inline SECURITY.md:6]: Which Istio version did you use and how was it installed? There are plenty of versions out there so let's add an example working config (so which kubectl apply command to do, or istioctl - that kinda business)

File: SECURITY.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+a way of introducing basic authentication for a certain domain and path.
+
+This assumes you have the dashboard exposed through the istio ingress on "example.com/build".
```
Comment (a-roberts) [inline SECURITY.md:9]: What steps were taken to get this please?

File: SECURITY.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    filterType: HTTP
+    filterName: envoy.lua
+    filterConfig:
+      inlineCode: |
+        local domain = "example.com"
+        local path = "/build/"
+
+        -- base64 encoded: <username>:<password>
+        -- example: ok:test2
+        local base64UsernamePassword = "Basic b2s6dGVzdDI="
```
Comment (a-roberts) [inline SECURITY.md:32]: So this is the value you'd change to set it to whatever you fancy right? Let's call that out with a helpful comment

**Comment by GijsvanDulmen** [comment]
> A few pointers to add please, looking good so far though and eager to give this a try (let us know what commands you've done and the working config so we've got a baseline anyone can follow along with, I think this'll be really useful)

@a-roberts I added some more information in the file. Could you review and let me know if anything more is needed?

**Comment by GijsvanDulmen** [comment]
> > > A few pointers to add please, looking good so far though and eager to give this a try (let us know what commands you've done and the working config so we've got a baseline anyone can follow along with, I think this'll be really useful)
> > 
> > 
> > @a-roberts I added some more information in the file. Could you review and let me know if anything more is needed?
> 
> That's perfect, thanks! I plan to try this Monday, my only suggestion now is to squash your commits into the one, the Prow set up for Tekton doesn't give us merge options where we could normally do this. Thanks again

Done.

**Comment by a-roberts** [comment]
Had a good try this today, set up Istio fine (1.5.2, demo profile following their docs) on Docker Desktop, applied the filter and the gateway. So for the gateway part, what do I have to do get this working on Docker Desktop (so, as much as "locally" and for development as possible please)? I've tried with 127.0.0.1, localhost, and keeping example.com. 

It'd be useful to see which environment you've got things working in too @GijsvanDulmen (so, what cloud in particular if it's not Docker Desktop) - if it's a big task for dev environments I'd be cool with mentioning this works on a particular set up only.

**Comment by GijsvanDulmen** [comment]
> Had a good try of this today, set up Istio fine (1.5.2, demo profile following their docs) on Docker Desktop, applied the filter and the gateway. So for the gateway part, what do I have to do get this working on Docker Desktop (so, as much as "locally" and for development as possible please)? I've tried with 127.0.0.1, localhost, and keeping example.com.
> 
> It'd be useful to see which environment you've got things working in too @GijsvanDulmen (so, what cloud in particular if it's not Docker Desktop) - if it's a big task for dev environments I'd be cool with mentioning this works on a particular set up only.

@a-roberts I got this working on a K8s 1.18.2 (4 node cluster, single master) setup with Istio 1.5.2. No Docker Desktop. I had another subdomain, something like "tools.example.com" (for obvious reasons hidden). This is a custom made cluster. I don't see a reason it wouldn't work with GKE however (I've got a similar setup on GKE running, however I haven't got the tekton dashboard running on it yet).

**Comment by a-roberts** [comment]
> > Had a good try of this today, set up Istio fine (1.5.2, demo profile following their docs) on Docker Desktop, applied the filter and the gateway. So for the gateway part, what do I have to do get this working on Docker Desktop (so, as much as "locally" and for development as possible please)? I've tried with 127.0.0.1, localhost, and keeping example.com.
> > It'd be useful to see which environment you've got things working in too @GijsvanDulmen (so, what cloud in particular if it's not Docker Desktop) - if it's a big task for dev environments I'd be cool with mentioning this works on a particular set up only.
> 
> @a-roberts I got this working on a K8s 1.18.2 (4 node cluster, single master) setup with Istio 1.5.2. No Docker Desktop. I had another subdomain, something like "tools.example.com" (for obvious reasons hidden). This is a custom made cluster. I don't see a reason it wouldn't work with GKE however (I've got a similar setup on GKE running, however I haven't got the tekton dashboard running on it yet).

Thanks for sharing, I think we're almost there then! Could you mention that version in the PR as well please and anything else you think might be useful, my fear is that we'll accept your helpful PR and you'll be the only one who can debug when things do wrong if there's not quite enough in there (as folks may try it on various other platforms).

Cheers

**Comment by GijsvanDulmen** [comment]
> > > Had a good try of this today, set up Istio fine (1.5.2, demo profile following their docs) on Docker Desktop, applied the filter and the gateway. So for the gateway part, what do I have to do get this working on Docker Desktop (so, as much as "locally" and for development as possible please)? I've tried with 127.0.0.1, localhost, and keeping example.com.
> > > It'd be useful to see which environment you've got things working in too @GijsvanDulmen (so, what cloud in particular if it's not Docker Desktop) - if it's a big task for dev environments I'd be cool with mentioning this works on a particular set up only.
> > 
> > 
> > @a-roberts I got this working on a K8s 1.18.2 (4 node cluster, single master) setup with Istio 1.5.2. No Docker Desktop. I had another subdomain, something like "tools.example.com" (for obvious reasons hidden). This is a custom made cluster. I don't see a reason it wouldn't work with GKE however (I've got a similar setup on GKE running, however I haven't got the tekton dashboard running on it yet).
> 
> Thanks for sharing, I think we're almost there then! Could you mention that version in the PR as well please and anything else you think might be useful, my fear is that we'll accept your helpful PR and you'll be the only one who can debug when things do wrong if there's not quite enough in there (as folks may try it on various other platforms). Even if it's mentioning this was done on a custom cluster - that's fine, provide as much as you can
> 
> Cheers

It's a rather "default" installation. So I added the versions above (sorry for the delay, started at a new job).

**Comment by eddycharly** [comment]
@GijsvanDulmen do you think it would be easy to write a walkthrough like this one https://github.com/tektoncd/dashboard/blob/master/docs/walkthrough/walkthrough-oauth2-proxy.md ?

It would be a great addition.
Would this work with GitHub authentication (or something else that an end user can setup easily) ?

**Comment by eddycharly** [comment]
@GijsvanDulmen i can't get the `EnvoyFilter` to work on my setup (kind + istio 1.6.7).
When looking at the istio docs, the `EnvoyFilter` CRD looks quite different than yours (https://istio.io/latest/docs/reference/config/networking/envoy-filter/).

`workloadLabels` vs `workloadSelector`, `filters` vs `configPatches`, etc... though both use `apiVersion: networking.istio.io/v1alpha3` and `kind: EnvoyFilter`.

Any idea what i should change to get this to work ?

**Comment by GijsvanDulmen** [comment]
> @GijsvanDulmen i can't get the `EnvoyFilter` to work on my setup (kind + istio 1.6.7).
> When looking at the istio docs, the `EnvoyFilter` CRD looks quite different than yours (https://istio.io/latest/docs/reference/config/networking/envoy-filter/).
> 
> `workloadLabels` vs `workloadSelector`, `filters` vs `configPatches`, etc... though both use `apiVersion: networking.istio.io/v1alpha3` and `kind: EnvoyFilter`.
> 
> Any idea what i should change to get this to work ?

Probably because mine was meant for istio 1.5? Could you provide the error message?

**Comment by eddycharly** [comment]
> Probably because mine was meant for istio 1.5? Could you provide the error message?

I don't have any error message, the filter creates correctly but never applies.

I'm not  sure this comes from the istio version, as the CRD GVK is the same, i can't imagine istio changing the CRD but keeping the same GVK, right ?

**Comment by eddycharly** [comment]
Even when looking at the docs for istio 1.5 https://istio.io/v1.5/docs/reference/config/networking/envoy-filter/ there's nothing like `spec.filters` or `spec.workloadLabels` in `EnvoyFilter`.

It looks like istio 1.2 (https://archive.istio.io/v1.2/docs/reference/config/networking/v1alpha3/envoy-filter/) but again it's `apiVersion: networking.istio.io/v1alpha3` and `kind: EnvoyFilter`.

**Comment by GijsvanDulmen** [comment]
> > Probably because mine was meant for istio 1.5? Could you provide the error message?
> 
> I don't have any error message, the filter creates correctly but never applies.
> 
> I'm not sure this comes from the istio version, as the CRD GVK is the same, i can't imagine istio changing the CRD but keeping the same GVK, right ?

Ok. I'm guessing the filter itself (the lua code) is impacted by the version 1.5 / 1.6 difference. You could try to add debug logging to the lua code to see if the code is even getting reached. I remember doing a lot of debugging code back when I made it to make sure it worked. Not really sure about the lua syntax ...

**Comment by eddycharly** [comment]
Ok, thanks. I definitely don't understand the CRD differences, i will try to ask on istio's slack to be sure.

**Comment by eddycharly** [comment]
I have adapted the filter and it's working now.
If you want to give a look #1677

**Pull request closed without merging**